File: ws/nacato/u2/ws23ExamPYandJS/ExamInterface.JS/controller.js
import { Bank } from './model.js';

export class BankController {
    constructor() {
        this.banks = [];

        this.entId = document.getElementById('ent_id');
        this.entName = document.getElementById('ent_name');
        this.entDeposits = document.getElementById('ent_deposits');
        this.entLoans = document.getElementById('ent_loans');
        
        this.btnCreate = document.getElementById('btn_create');
        this.btnUpdate = document.getElementById('btn_update');
        this.btnDelete = document.getElementById('btn_delete');
        this.btnClear = document.getElementById('btn_clear');
        this.tableBody = document.getElementById('table_body');

        this.btnCreate.onclick = () => this.createBank();
        this.btnUpdate.onclick = () => this.updateBank();
        this.btnDelete.onclick = () => this.deleteBank();
        this.btnClear.onclick = () => this.clearInputs();
        
        this.updateTable();
    }

    getBankData() {
        const id = this.entId.value.trim();
        const name = this.entName.value.trim();
        const depositsRaw = this.entDeposits.value;
        const loansRaw = this.entLoans.value;

        if (!id || !name || depositsRaw === "" || loansRaw === "") {
            throw new Error("All fields are required.");
        }

        const total_deposits = parseFloat(depositsRaw);
        const total_loans = parseFloat(loansRaw);

        if (isNaN(total_deposits) || isNaN(total_loans)) {
            throw new Error("Deposits and Loans must be valid numbers.");
        }

        return { id, name, total_deposits, total_loans };
    }

    createBank() {
        try {
            const data = this.getBankData();

            if (this.banks.some(b => b.id === data.id)) {
                alert("Error: A bank with this ID already exists.");
                return;
            }

            const newBank = new Bank(data.id, data.name, data.total_deposits, data.total_loans);
            this.banks.push(newBank);

            alert(`Bank '${newBank.name}' added successfully.`);
            this.clearInputs();
            this.updateTable(); 
        } catch (error) {
            alert(error.message);
        }
    }

    updateBank() {
        try {
            const data = this.getBankData();
            const index = this.banks.findIndex(b => b.id === data.id);

            if (index !== -1) {
                this.banks[index].name = data.name;
                this.banks[index].total_deposits = data.total_deposits;
                this.banks[index].total_loans = data.total_loans;

                alert("Bank updated successfully.");
                this.clearInputs();
                this.updateTable(); 
            } else {
                alert("Error: Bank ID not found. You cannot change the ID.");
            }
        } catch (error) {
            alert(error.message);
        }
    }

    deleteBank() {
        const idToDelete = this.entId.value.trim();

        if (!idToDelete) {
            alert("Please enter or select a Bank ID to delete.");
            return;
        }

        const index = this.banks.findIndex(b => b.id === idToDelete);

        if (index !== -1) {
            this.banks.splice(index, 1);
            alert("Bank deleted successfully.");
            this.clearInputs();
            this.updateTable(); 
        } else {
            alert("Error: Bank ID not found.");
        }
    }

    clearInputs() {
        this.entId.value = "";
        this.entName.value = "";
        this.entDeposits.value = "";
        this.entLoans.value = "";
    }

    updateTable() {
        this.tableBody.innerHTML = "";

        if (this.banks.length === 0) {
            this.tableBody.innerHTML = `<tr><td colspan="5" style="text-align:center; color:gray;">No registered banks.</td></tr>`;
            return;
        }

        this.banks.forEach(bank => {
            const row = document.createElement('tr');
            
            row.innerHTML = `
                <td>${bank.id}</td>
                <td>${bank.name}</td>
                <td>$${bank.total_deposits.toFixed(2)}</td>
                <td>$${bank.total_loans.toFixed(2)}</td>
                <td>$${bank.computeNetWorth().toFixed(2)}</td>
            `;

            row.onclick = () => {
                this.entId.value = bank.id;
                this.entName.value = bank.name;
                this.entDeposits.value = bank.total_deposits;
                this.entLoans.value = bank.total_loans;
            };

            this.tableBody.appendChild(row);
        });
    }
}

File: ws/nacato/u2/ws23ExamPYandJS/ExamInterface.PY/Exam2PY/controller.py
from model import Bank
from view import BankView

class BankController:
    def __init__(self, view: BankView):
        self.banks = []
        self.view = view
        
        self.view.btn_create.config(command=self.create_bank)
        self.view.btn_update.config(command=self.update_bank)
        self.view.btn_delete.config(command=self.delete_bank)
        self.view.btn_clear.config(command=self.view.clear_inputs)
        
        self.view.tree.bind("<<TreeviewSelect>>", self.on_tree_select)

    def create_bank(self):
        try:
            data = self.view.get_bank_data()
            if not data["id"] or not data["name"]:
                self.view.display_error("All fields are required.")
                return

            for bank in self.banks:
                if bank.id == data["id"]:
                    self.view.display_error("A bank with this ID already exists.")
                    return

            new_bank = Bank(data["id"], data["name"], data["total_deposits"], data["total_loans"])
            self.banks.append(new_bank)
            
            self.view.display_message(f"Bank '{new_bank.name}' added successfully.")
            self.view.clear_inputs()
            self.view.update_tree(self.banks)
        except ValueError as e:
            self.view.display_error(str(e))

    def update_bank(self):
        try:
            data = self.view.get_bank_data()
            
            for bank in self.banks:
                if bank.id == data["id"]:
                    bank.name = data["name"]
                    bank.total_deposits = data["total_deposits"]
                    bank.total_loans = data["total_loans"]
                    
                    self.view.display_message("Bank updated successfully.")
                    self.view.clear_inputs()
                    self.view.update_tree(self.banks)
                    return
            
            self.view.display_error("Bank ID not found. You cannot change the ID of a bank.")
        except ValueError as e:
            self.view.display_error(str(e))

    def delete_bank(self):
        id_to_delete = self.view.ent_id.get().strip()
        
        if not id_to_delete:
            self.view.display_error("Please enter or select a Bank ID to delete.")
            return

        for bank in self.banks:
            if bank.id == id_to_delete:
                self.banks.remove(bank)
                self.view.display_message("Bank deleted successfully.")
                self.view.clear_inputs()
                self.view.update_tree(self.banks)
                return
                
        self.view.display_error("Bank ID not found.")

    def on_tree_select(self, event):
        selected_items = self.view.tree.selection()
        if not selected_items:
            return
            
        item = selected_items[0]
        values = self.view.tree.item(item, "values")
        
        id_val = values[0]
        name_val = values[1]
        deposits_val = values[2].replace("$", "")
        loans_val = values[3].replace("$", "")
        
        self.view.set_bank_data(id_val, name_val, deposits_val, loans_val)

    def start(self):
        self.view.mainloop()